htlcswitch/switch.go

```diff
@@ -0,0 +1,770 @@
+// NOTE: Should be run as goroutine.
+func (s *Switch) startHandling() {
+	defer s.wg.Done()
+
+	// TODO(roasbeef): cleared vs settled distinction
+	var prevNumUpdates uint64
+	var prevSatSent btcutil.Amount
+	var prevSatRecv btcutil.Amount
+
+	for {
+		select {
+		case req := <-s.chanCloseRequests:
+			s.handleChanelClose(req)
+
+		case cmd := <-s.forwardCommands:
+			var paymentHash lnwallet.PaymentHash
+			var amount btcutil.Amount
+
+			switch m := cmd.pkt.htlc.(type) {
+			case *lnwire.UpdateAddHTLC:
+				paymentHash = m.PaymentHash
+				amount = m.Amount
+			case *lnwire.UpdateFufillHTLC, *lnwire.UpdateFailHTLC:
+				paymentHash = cmd.pkt.payHash
+				amount = cmd.pkt.amount
+			default:
+				cmd.err <- errors.New("wrong type of update")
+				return
+			}
+
+			payment, err := s.findPayment(amount, paymentHash)
+			if err != nil {
+				cmd.err <- s.handlePacketForward(cmd.pkt)
+			} else {
+				cmd.err <- s.handleLocalDispatch(payment, cmd.pkt)
+			}
+
+		case <-time.Tick(10 * time.Second):
+			var overallNumUpdates uint64
+			var overallSatSent btcutil.Amount
+			var overallSatRecv btcutil.Amount
+
+			for _, link := range s.links {
+				updates, sent, recv := link.Stats()
+				overallNumUpdates += updates
+				overallSatSent += sent
+				overallSatRecv += recv
+			}
+
+			if overallNumUpdates == 0 {
+				continue
+			}
+
+			diffNumUpdates := overallNumUpdates - prevNumUpdates
+			diffSatSent := overallSatSent - prevSatSent
+			diffSatRecv := overallSatRecv - prevSatRecv
+
+			log.Infof("sent %v satoshis received %v satoshi "+
+				" in the last 10 seconds (%v tx/sec)",
+				diffSatSent, diffSatRecv, float64(diffNumUpdates)/10)
+
+			prevNumUpdates = overallNumUpdates
+			prevSatSent = overallSatSent
+			prevSatRecv = overallSatRecv
+
+		case <-s.quit:
+			return
+		}
+	}
+}
+
+// Start starts all helper goroutines required for the operation of the switch.
+func (s *Switch) Start() error {
+	if !atomic.CompareAndSwapInt32(&s.started, 0, 1) {
+		log.Warn("Htlc Switch already started")
+		return nil
+	}
+
+	log.Infof("Htlc Switch starting")
+
+	s.wg.Add(1)
+	go s.startHandling()
+
+	return nil
+}
+
+// Stop gracefully stops all active helper goroutines, then waits until they've
+// exited.
+func (s *Switch) Stop() error {
+	if !atomic.CompareAndSwapInt32(&s.shutdown, 0, 1) {
+		log.Warn("Htlc Switch already stopped")
+		return nil
+	}
+
+	log.Infof("Htlc Switch shutting down")
+
+	s.mutex.Lock()
+	for _, link := range s.links {
+		delete(s.links, link.ChanID())
+		go link.Stop()
+		log.Infof("Remove channel link with ChannelID(%v)", link.ChanID())
+	}
+	s.mutex.Unlock()
+
+	close(s.quit)
+	s.wg.Wait()
+
+	return nil
+}
+
+// AddLink is used to add and start the newly created channel link and start
+// use it to handle the channel updates.
+func (s *Switch) AddLink(link ChannelLink) error {
+	if err := link.Start(); err != nil {
+		return err
+	}
+
+	s.mutex.Lock()
+	s.links[link.ChanID()] = newBoundedLinkChan(numSlots, link)
```

Comment:
need to be set _beofre_ starting

The link should be added to the set of active links _before_ the link has been started. Otherwise, it's possible that a link message is sent before it's fully registered within the switch.